src/Compilers/Core/MSBuildTask/MapSourceRoots.cs

```diff
@@ -0,0 +1,232 @@
+        /// </summary>
+        [Required]
+        public ITaskItem[] SourceRoots { get; set; }
+
+        /// <summary>
+        /// True if the mapped paths should be deterministic.
+        /// </summary>
+        public bool Deterministic { get; set; }
+
+        /// <summary>
+        /// SourceRoot items with <term>MappedPath</term> metadata set.
+        /// Items listed in <see cref="SourceRoots"/> that have the same ItemSpec will be merged into a single item in this list.
+        /// </summary>
+        [Output]
+        public ITaskItem[] MappedSourceRoots { get; private set; }
+
+        private static class Names
+        {
+            public const string SourceRoot = nameof(SourceRoot);
+            public const string DeterministicSourcePaths = nameof(DeterministicSourcePaths);
+
+            // Names of well-known SourceRoot metadata items:
+            public const string SourceControl = nameof(SourceControl);
+            public const string RevisionId = nameof(RevisionId);
+            public const string NestedRoot = nameof(NestedRoot);
+            public const string ContainingRoot = nameof(ContainingRoot);
+            public const string MappedPath = nameof(MappedPath);
+            public const string SourceLinkUrl = nameof(SourceLinkUrl);
+
+            public static readonly string[] SourceRootMetadataNames = new[] { SourceControl, RevisionId, NestedRoot, ContainingRoot, MappedPath, SourceLinkUrl };
+        }
+
+        private static string EndWithSlash(string path)
+            => (path[path.Length - 1] == '/') ? path : path + '/';
+
+        private static bool EndsWithDirectorySeparator(string path)
+        {
+            if (path.Length == 0)
+            {
+                return false;
+            }
+
+            char c = path[path.Length - 1];
+            return c == Path.DirectorySeparatorChar || c == Path.AltDirectorySeparatorChar;
+        }
+
+        private void MergeSourceRootMetadata(ITaskItem left, ITaskItem right)
+        {
+            foreach (string metadataName in right.MetadataNames)
+            {
+                var leftValue = left.GetMetadata(metadataName);
+                var rightValue = right.GetMetadata(metadataName);
+
+                if (!string.IsNullOrEmpty(leftValue) && !string.IsNullOrEmpty(rightValue))
+                {
+                    Log.LogErrorFromResources("MapSourceRoots.ContainsDuplicate", Names.SourceRoot, right.ItemSpec, metadataName, leftValue, rightValue);
+                }
+                else if (string.IsNullOrEmpty(leftValue) && !string.IsNullOrEmpty(rightValue))
+                {
+                    left.SetMetadata(metadataName, rightValue);
+                }
+            }
+        }
+
+        public override bool Execute()
+        {
+            // Merge metadata of SourceRoot items with the same identity.
+            var mappedSourceRoots = new List<ITaskItem>();
+            var rootByItemSpec = new Dictionary<string, ITaskItem>();
```

Comment:
>           var rootByItemSpec = new Dictionary<string, ITaskItem>(); [](start = 1, length = 68)

Why are we using ordinal path comparison here? The compiler is pretty inconsistent here but generally favors insensiitive.

Reply:
Are you saying we do not support source directories that differ by casing on Linux?

Reply:
Different parts of the compiler differ here. It's not consistent.